docs/language-reference/04-types.md

```diff
@@ -0,0 +1,323 @@
+Thus the preceding declaration is equivalent to:
+
+```hlsl
+int a[4] = { 0xA, 0xB, 0xC, 0xD };
+```
+
+A variable declared in this fashion semantically has a known-size array type and not an unknown-size array type; the use of an unknown-size array type for the declaration is just a convenience feature.
+
+### Standard Layout
+
+The _stride_ of a type is the smallest multiple of its alignment not less than its size.
+
+Using the standard layout for an array type `T[]` or `T[N]`:
+
+* The _element stride_ of the array type is the stride of its element type `T`
+* Element `i` of the array starts at an offset that is `i` times the element stride of the array
+* The alignment of the array type is the alignment of `T`
+* The size of an unknown-size array type is unknown
+* The size of a known-size array with zero elements is zero
+* The size of a known-size array with a nonzero number `N` of elements is the size of `T` plus `N - 1` times the element stride of the array
+
+### C-Style Layout
+
+The C-style layout of an array type differs from the standard layout in that the size of a known-size array with a nonzero number `N` of elements is `N` times the element stride of the array.
+
+### D3D Constant Buffer Layout
+
+The D3D constant buffer layout of an array differs from the standard layout in that the element stride of the array is set to the smallest multiple of the alignment of `T` that is not less than the stride of `T`
+
+This Type
+---------
+
+Within the body of a structure or interface declaration, the keyword `This` may be used to refer to the enclosing type.
+Inside of a structure type declaration, `This` refers to the structure type itself.
+Inside of an interface declaration, `This` refers to the concrete type that is conforming to the interface (that is, the type of `this`).
+
+Opaque Types
+------------
+
+_Opaque_ types are built-in types that (depending on the target platform) may not have a well-defined size or representation in memory.
+Similar languages may refer to these as "resource types" or "object types."
+
+The full list of opaque types supported by Slang can be found in the standard library reference, but important examples are:
+
+* Texture types such as `Texture2D<T>`, `TextureCubeArray<T>`, and `RWTexture2DMS<T>`
+* Sampler state types: `SamplerState` and `SamplerComparisonState`
+* Buffer types like `ConstantBuffer<T>` and  `StructuredBuffer<T>`
+* Parameter blocks: `ParameterBlock<T>`
+
+Layout for opaque types depends on the target platform, and no specific guarantees can be made about layout rules across platforms.
+
+Known and Unknown Size
+----------------------
+
+Every type has either known or unknown size.
+Types with unknown size arise in a few ways:
+
+* An unknown-size array type has unknown size
+
+* A structure type has unknown size if any field type has unknown size
+
+The use of tyeps with unknown size is restricted as follows:
+
+* A type with unknown size cannot be used as the element type of an array
```

Comment:
No problem parsing it, but it is semantically meaningless to our layout model (so it *should* be illegal).

From what I remember, in C/C++ if the element type of an array is itself an unknown-size array then the latter array "decays" into a pointer type, so that the result is equivalent to defining an array of pointers, which is of course legal, but isn't what a Slang array is meant to represent.